core/Enclave/enclave_hsm.cpp

```diff
@@ -940,4 +940,73 @@ sgx_status_t enclave_verify_quote_policy(uint8_t *quote, uint32_t quote_size,
         return SGX_ERROR_UNEXPECTED;
     }
     return SGX_SUCCESS;
+}
+
+/**
+ * @brief Generate HMAC with given cmk, apikey and payload
+ * @param cmk the cmk for apikey decryption
+ * @param apikey the encrypted apikey
+ * @param payload the payload of HMAC
+ * @param hmac the output of the function
+*/
+sgx_status_t enclave_generate_hmac(ehsm_keyblob_t *cmk, uint32_t cmk_size,
+                                    ehsm_data_t *apikey, uint32_t apikey_size,
+                                    ehsm_data_t *payload, uint32_t payload_size,
+                                    ehsm_data_t *hmac, uint32_t hmac_size)
+{
+    if (cmk == NULL ||
+        cmk_size != APPEND_SIZE_TO_KEYBLOB_T(cmk->keybloblen) ||
+        cmk->keybloblen == 0 ||
+        cmk->metadata.origin != EH_INTERNAL_KEY ||
+        cmk->metadata.keyusage != EH_KEYUSAGE_ENCRYPT_DECRYPT)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    if (apikey == NULL ||
+        apikey_size != APPEND_SIZE_TO_DATA_T(apikey->datalen))
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    if (payload == NULL ||
+        payload_size != APPEND_SIZE_TO_DATA_T(payload->datalen) ||
+        payload->datalen == 0)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    if (hmac == NULL ||
+        hmac_size != APPEND_SIZE_TO_DATA_T(hmac->datalen) ||
+        hmac->datalen != EH_HMAC_SHA256_SIZE)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+    ehsm_data_t *aad = NULL;
+    ehsm_data_t *rawApiKey = NULL;
+
+    // create empty aad
+    aad = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(0));
+    if (aad == NULL)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    aad->datalen = 0;
+
+    // create space for storing raw apikey
+    rawApiKey = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(EH_API_KEY_SIZE));
+    if (rawApiKey == NULL)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    rawApiKey->datalen = EH_API_KEY_SIZE;
+
+    // 1. Decrypt the apikey
+    ret = enclave_decrypt(cmk, cmk_size, aad, APPEND_SIZE_TO_DATA_T(aad->datalen), apikey, apikey_size, rawApiKey, APPEND_SIZE_TO_DATA_T(rawApiKey->datalen));
+    if (ret != SGX_SUCCESS) {
+        log_w("apikey decrypt failed");
+        goto out;
+    }
+
+    // 2. Generate HMAC
+    ret = sgx_hmac_sha256_msg(payload->data, payload->datalen, rawApiKey->data, rawApiKey->datalen, hmac->data, hmac->datalen);
+
+out:
+    SAFE_FREE(aad);
+    SAFE_FREE(rawApiKey);
```

Comment:
before free the sensitive data, better to clear it by memset_s(); 
you can refer to the samples in this same file.

Reply:
Thanks for reviewing! The sensitive info (`rawApiKey` in this function) should be set to 0 before returning, please check.